hax-lib-protocol/src/lib.rs

```diff
@@ -0,0 +1,24 @@
+#[derive(Debug)]
+pub enum ProtocolError {
+    InvalidMessage,
+    InvalidPrologue,
+}
+
+pub type ProtocolResult<T> = Result<T, ProtocolError>;
+
+pub trait InitialState {
+    fn init(prologue: Option<Vec<u8>>) -> ProtocolResult<Self>
+    where
+        Self: Sized;
+}
+
+pub trait WriteState {
+    type NextState;
+    type Message;
+    fn write(self) -> ProtocolResult<(Self::NextState, Self::Message)>;
+}
+
+pub trait ReadState<NextState> {
+    type Message;
```

Comment:
I chose a generic here, because a state might implement `ReadState<NextState>` for different successor states, depending on the message that is read. If the next state was an associated type, there could only ever be one possible next state, since the trait could only be implemented once, no? On the flip side, I've made the assumption that in a `WriteState` the message to write and the resulting next state are always uniquely determined.

Reply:
Agreed, makes sense. But because it's different in `WriteState` this needs explanation. (Everything in here needs documentation, but something like this in particular.)

Reply:
Added it to the documentation.